### PR TITLE
AP_HAL_ESP32: update Util library to work with ToneAlarm

### PR DESCRIPTION
**Pull request by radityankn**
implemented tonealarm init and set buzzer frequency function for ESP32
to use ToneAlarm, define HAL_NOTIFY_TONEALARM_ENABLED TRUE and 
HAL_PWM_BUZZER_PIN to corresponding ESP32 pin

### CODE REVIEW

File: libraries/AP_HAL_ESP32/Util.cpp [hunk 1]
```diff
@@ -153,26 +153,50 @@ Util::safety_state Util::safety_switch_state(void)
...
 void Util::toneAlarm_set_buzzer_tone(float frequency, float volume, uint32_t duration_ms)
 {
-    if (is_zero(frequency) || is_zero(volume)) {
-        pwmDisableChannel(_toneAlarm_pwm_group.pwm_drv, _toneAlarm_pwm_group.chan);
-    } else {
-        pwmChangePeriod(_toneAlarm_pwm_group.pwm_drv,
-                        roundf(_toneAlarm_pwm_group.pwm_cfg.frequency/frequency));
-
-        pwmEnableChannel(_toneAlarm_pwm_group.pwm_drv, _toneAlarm_pwm_group.chan, roundf(volume*_toneAlarm_pwm_group.pwm_cfg.frequency/frequency)/2);
+    if (frequency == 0){
```
Comment (muramura) [inline libraries/AP_HAL_ESP32/Util.cpp:193]: Match the code style with the existing.

File: libraries/AP_HAL_ESP32/Util.cpp [hunk 2]
```diff
@@ -153,26 +153,50 @@ Util::safety_state Util::safety_switch_state(void)
...
+    	ledc_stop(_toneAlarm_pwm_group.channel_config.speed_mode, _toneAlarm_pwm_group.channel_config.channel, 0);
+    }
+    else {
```
Comment (muramura) [inline libraries/AP_HAL_ESP32/Util.cpp:196]: Match the code style with the existing.

File: libraries/AP_HAL_ESP32/Util.h [hunk 3]
```diff
@@ -54,9 +58,10 @@ class ESP32::Util : public AP_HAL::Util
     bool get_system_id(char buf[50]) override;
     bool get_system_id_unformatted(uint8_t buf[], uint8_t &len) override;
 
-#ifdef HAL_PWM_ALARM
-    bool toneAlarm_init() override;
+#if AP_NOTIFY_TONEALARM_ENABLED
+    bool toneAlarm_init(uint8_t types) override;
     void toneAlarm_set_buzzer_tone(float frequency, float volume, uint32_t duration_ms) override;
+    static uint8_t _toneAlarm_types;
```
Comment (muramura) [inline libraries/AP_HAL_ESP32/Util.h:64]: I think this variable should be a PRIVATE variable and the STATIC designation is unnecessary.

**Comment by davidbuzz** [comment]
I'm generally ok with this in-principal, using the ledc driver for the buzzer is a clean idea.
Please fix the minor styling issues and rebase , and run fix one of the commit messages to it has the AP_HAL_ESP32: prefix and it can go in.

**Comment by davidbuzz** [comment]
needs rebasing and re-testing since the esp32 has now moved to IDF 5

**Pull request closed without merging**